File: my-folder/0061-rotate-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int findSize(ListNode* head){
        int len = 0;
        while(head){
            len++;
            head=head->next;
        }
        return len;
    }
    
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head || k == 0)
            return head;
        
        int n = findSize(head);
        k = k % n;
        if (k == 0) 
            return head;

        ListNode* l1 = head;
        int i = 0;
        while(i < n - k-1){
            l1 = l1->next;
            i++;
        }
        ListNode* l2 = l1->next;
        l1->next = nullptr;
        ListNode* l2h = l2;
        while(l2->next){
            l2 = l2->next;
        }
        l2->next = head;
        return l2h;
    }
};

File: my-folder/0061-rotate-list/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode reverse(ListNode head){
        ListNode prev = null;
        while(head!=null){
            ListNode temp = head.next;
            head.next = prev;
            prev=head;
            head=temp;
        }
        return prev;
    }
    int lenList(ListNode head){
        int len=0;
        while(head!=null){
            head=head.next;
            len++;
        }
        return len;
    }

    public ListNode rotateRight(ListNode head, int k) {
        int length = lenList(head);
        if(length==0) return null;
        k=k%length;
        if(k==0) return head;
        ListNode newH = reverse(head);
        ListNode dummy = new ListNode(-1);
        dummy.next = newH;
        ListNode start = newH;
        
        for(int i=1;i<k;i++){
            newH = newH.next;
        }
        if(newH==null) return dummy.next;
        ListNode temp = newH.next;
        newH.next = null;
        dummy.next=reverse(start);
        start.next=reverse(temp);
        return dummy.next;


        
    }
}
